testsuite/tests/typing-jkind-bounds/subsumption/recursive.ml

```diff
@@ -0,0 +1,223 @@
+(* TEST
+    flags = "-extension layouts_alpha";
+    expect;
+*)
+
+type 'a my_list : immutable_data with 'a = Nil | Cons of 'a * 'a my_list
+[%%expect {|
+type 'a my_list = Nil | Cons of 'a * 'a my_list
+|}]
+
+type 'a my_list : immutable_data with 'a = 'a list = [] | ( :: ) of 'a * 'a my_list
+[%%expect {|
+type 'a my_list = 'a list = [] | (::) of 'a * 'a my_list
+|}]
+
+type 'a my_list : immutable_data with 'a = Nil | Cons of 'a * 'a foo
+and 'a foo = 'a my_list
+[%%expect {|
+type 'a my_list = Nil | Cons of 'a * 'a foo
+and 'a foo = 'a my_list
+|}]
+
+(* CR layouts v2.8: this should be accepted *)
+type 'a my_list : immutable_data with 'a =
+  | Nil
+  | Cons of 'a * 'a my_list my_list my_list my_list my_list my_list my_list my_list my_list my_list my_list my_list
+[%%expect {|
+Line 3, characters 17-115:
+3 |   | Cons of 'a * 'a my_list my_list my_list my_list my_list my_list my_list my_list my_list my_list my_list my_list
+                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: Layout mismatch in final type declaration consistency check.
+       This is most often caused by the fact that type inference is not
+       clever enough to propagate layouts through variables in different
+       declarations. It is also not clever enough to produce a good error
+       message, so we'll say this instead:
+         The layout of 'a my_list my_list my_list my_list my_list my_list
+                       my_list my_list my_list my_list my_list is any
+           because the .cmi file for my_list is missing.
```

Comment:
Zoiks. This is pretty awful.

Reply:
Yeah. I'm tempted to make error messages say something about running out of fuel. But I think that's outside the scope of this PR. I'll leave a CR here about it